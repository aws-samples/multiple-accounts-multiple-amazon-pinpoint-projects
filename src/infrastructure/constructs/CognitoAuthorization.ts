/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {Construct, Dependable, IDependable} from "constructs";
import {AdvancedSecurityMode, CfnUserPoolGroup, CfnUserPoolUser, CfnUserPoolUserToGroupAttachment, OAuthScope, ResourceServerScope, UserPool, UserPoolDomain} from "aws-cdk-lib/aws-cognito";
import {AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId} from "aws-cdk-lib/custom-resources";
import {Aws, CfnOutput, Duration} from "aws-cdk-lib";
import {CustomDomainOptions} from "aws-cdk-lib/aws-cognito/lib/user-pool-domain";
import {CnameRecord, IHostedZone} from "aws-cdk-lib/aws-route53";


export interface CognitoAuthorizationConfig {
	customDomain?: CustomDomainOptions;
	hostedZone?: IHostedZone;
	apiTempPassword?: string

}

export class CognitoAuthorization extends Construct implements IDependable {
	readonly userPool: UserPool;
	readonly customUserPoolDomain: UserPoolDomain | undefined;

	constructor(scope: Construct, id: string, config: CognitoAuthorizationConfig) {
		super(scope, id);
		Dependable.implement(this, {
			dependencyRoots: [this],
		});
		this.node.addValidation({
			validate(): string[] {
				const messages: string[] = [];
				if (config.customDomain != undefined && config.hostedZone == undefined) {
					messages.push("You must specify hostedZone if customDomain is set")
				}
				if (config.customDomain != undefined && config.hostedZone == undefined) {
					messages.push("You must specify customDomain if hostedZone is set")
				}

				if (config.apiTempPassword != undefined && !new RegExp("^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\\$%\\^&\\*])(.{9,})$").test(config.apiTempPassword)) {
					messages.push("The temp password for the API admin user must be at least 9 characters long, include digits, both lower and uppercase letters as well as symbols.")
				}

				return messages;
			},
		});
		const userPoolName = 'pinpoint-management-account-user-pool'
		this.userPool = new UserPool(this, 'UserPool', {
			userPoolName: userPoolName,
			advancedSecurityMode: AdvancedSecurityMode.ENFORCED,
			passwordPolicy: {
				minLength: 9,
				requireDigits: true,
				requireLowercase: true,
				requireSymbols: true,
				requireUppercase: true,
				tempPasswordValidity: Duration.days(1)
			}
		});


		const adminScope = new ResourceServerScope({
			scopeName: "Admin",
			scopeDescription: "Admin Scope"
		})
		const adminResourceServer = this.userPool.addResourceServer("ApiResourceServer", {
			identifier: "api",
			scopes: [adminScope],
			userPoolResourceServerName: "ApiResourceServer"

		})
		const implicitClient = this.userPool.addClient('ImplicitClient', {
			authFlows: {
				userSrp: false,
				userPassword: true,
				adminUserPassword: false,
				custom: false
			},
			userPoolClientName: `${userPoolName}-implicit-client`,
			oAuth: {
				flows: {
					implicitCodeGrant: true,
					clientCredentials: false,
					authorizationCodeGrant: false,
				},
				scopes: [OAuthScope.resourceServer(adminResourceServer, adminScope)],
				callbackUrls: ['https://oauth.pstmn.io/v1/callback'],

			},
		});
		const clientCredentialsClient = this.userPool.addClient('ClientCredentialsClient', {
			authFlows: {
				userSrp: false,
				userPassword: true,
				adminUserPassword: false,
				custom: false
			},
			generateSecret: true,
			userPoolClientName: `${userPoolName}-client-credentials-client`,
			oAuth: {
				flows: {
					implicitCodeGrant: false,
					clientCredentials: true,
					authorizationCodeGrant: false,
				},
				scopes: [OAuthScope.resourceServer(adminResourceServer, adminScope)],
				callbackUrls: ['https://oauth.pstmn.io/v1/callback'],

			},
		});
		const adminUserGroupAttachment = new CfnUserPoolUserToGroupAttachment(this, 'AdminUserGroupAttachment', {
			groupName: 'Admin',
			username: 'admin',
			userPoolId: this.userPool.userPoolId,
		});
		const adminUser = new CfnUserPoolUser(this, 'AdminUser', {
			userPoolId: this.userPool.userPoolId,
			username: 'admin',
		});
		const adminGroup = new CfnUserPoolGroup(this, 'AdminGroup', {
			userPoolId: this.userPool.userPoolId,
			description: 'Administrator',
			groupName: 'Admin',
		});


		adminUserGroupAttachment.addDependency(adminUser)
		adminUserGroupAttachment.addDependency(adminGroup)
		if (config.apiTempPassword != undefined) {
			const adminSetPassword = new AwsCustomResource(this, 'AdminSetUserPassword', {
				onCreate: {
					service: 'CognitoIdentityServiceProvider',
					action: 'adminSetUserPassword',
					parameters: {
						UserPoolId: this.userPool.userPoolId,
						Username: 'admin',
						Password: config.apiTempPassword,
						Permanent: false,
					},
					physicalResourceId: PhysicalResourceId.of('AdminSetUserPassword'),
				},
				policy: AwsCustomResourcePolicy.fromSdkCalls({resources: AwsCustomResourcePolicy.ANY_RESOURCE}),
				installLatestAwsSdk: true,
			});
			adminSetPassword.node.addDependency(adminUser);
			new CfnOutput(this, "ApiAdminUsernameOutput", {
				key: "ApiAdminUsername",
				value: "admin"
			})
			new CfnOutput(this, "ApiAdminTempPasswordOutput", {
				key: "ApiAdminTempPassword",
				value: config.apiTempPassword
			})
		}
		this.userPool.addDomain('CognitoDomain', {
			cognitoDomain: {
				domainPrefix: `sms-${Aws.ACCOUNT_ID}`
			}
		});


		if (config.customDomain != undefined && config.hostedZone != undefined) {
			this.customUserPoolDomain = this.userPool.addDomain('CustomDomain', {
				customDomain: config.customDomain
			});
			const cname = new CnameRecord(this, "AuthCnameRecord", {
				recordName: config.customDomain.domainName,
				zone: config.hostedZone,
				domainName: this.customUserPoolDomain.cloudFrontDomainName
			})
			cname.node.addDependency(this.customUserPoolDomain)
			new CfnOutput(this, "UserPoolDomainBaseUrlOutput", {
				key: "UserPoolDomainBaseUrl",
				value: this.customUserPoolDomain.baseUrl()
			})
			new CfnOutput(this, "ImplicitClientLoginUrlOutput", {
				key: "ImplicitClientLoginUrl",
				value: this.customUserPoolDomain.signInUrl(implicitClient, {redirectUri: "https://oauth.pstmn.io/v1/callback"})
			})
		}

		new CfnOutput(this, "ImplicitClientIdOutput", {
			key: "ImplicitClientId",
			value: implicitClient.userPoolClientId
		})
		new CfnOutput(this, "ClientCredentialsClientIdOutput", {
			key: "ClientCredentialsClientId",
			value: clientCredentialsClient.userPoolClientId
		})


	}
}