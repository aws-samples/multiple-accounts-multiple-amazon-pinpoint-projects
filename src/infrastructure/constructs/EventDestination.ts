/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import path from "path";
import { IDeliveryStream } from "@aws-cdk/aws-kinesisfirehose-alpha";
import { EventType } from "@aws-sdk/client-pinpoint-sms-voice-v2";
import { Aws, CustomResource, Duration, RemovalPolicy } from "aws-cdk-lib";
import { Effect, IRole, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Runtime, Tracing } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";

import { ITopic } from "aws-cdk-lib/aws-sns";
import { Construct, Dependable, IDependable } from "constructs";

export interface EventDestinationConfig {
  configurationSetName: string; // required
  eventDestinationName: string; // required
  matchingEventTypes: EventType[];
}

export interface CloudWatchLogsEventDestinationConfig
  extends EventDestinationConfig {
  role: IRole;
  logGroup: LogGroup;
}

export interface KinesisFirehoseEventDestinationConfig
  extends EventDestinationConfig {
  streamWriterRole: IRole;
  destination: IDeliveryStream;
  bucketWriterRole: IRole;
}

export interface SnsEventDestinationConfig extends EventDestinationConfig {
  topic: ITopic;
}

export abstract class EventDestination
  extends Construct
  implements IDependable
{
  readonly name: string;
  protected customResourceFn: NodejsFunction;

  protected constructor(
    scope: Construct,
    id: string,
    properties: { [key: string]: any },
  ) {
    super(scope, id);
    Dependable.implement(this, {
      dependencyRoots: [this],
    });
    this.node.addValidation({
      validate(): string[] {
        const messages: string[] = [];

        return messages;
      },
    });

    const logGroup = new LogGroup(
      this,
      "EventDestinationCustomResourceFnLogGroup",
      {
        retention: RetentionDays.ONE_MONTH,
      },
    );
    this.customResourceFn = new NodejsFunction(
      this,
      "EventDestinationCustomResourceFn",
      {
        description: "EventDestination CustomResource Function",
        memorySize: 256,
        timeout: Duration.seconds(30),
        runtime: Runtime.NODEJS_LATEST,
        handler: "index.onEvent",
        entry: path.join(
          __dirname,
          "..",
          "..",
          "runtime",
          "EventDestination.ts",
        ),
        logGroup: logGroup,

        environment: {
          LOG_LEVEL: "DEBUG",
        },
        tracing: Tracing.ACTIVE,
        initialPolicy: [
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["sms-voice:*EventDestination"],
            resources: [
              `arn:${Aws.PARTITION}:sms-voice:${Aws.REGION}:${Aws.ACCOUNT_ID}:configuration-set/*`,
            ],
          }),
        ],
      },
    );
    new CustomResource(this, "EventDestinationCustomResource", {
      serviceToken: this.customResourceFn.functionArn,
      removalPolicy: RemovalPolicy.DESTROY,
      properties: properties,
    });
    this.name = properties.EventDestinationName;
  }
}

export class CloudWatchLogsEventDestination extends EventDestination {
  constructor(
    scope: Construct,
    id: string,
    config: CloudWatchLogsEventDestinationConfig,
  ) {
    super(scope, id, {
      ConfigurationSetName: config.configurationSetName, // required
      EventDestinationName: config.eventDestinationName, // required
      MatchingEventTypes: config.matchingEventTypes,
      CloudWatchLogsDestination: {
        // CloudWatchLogsDestination
        IamRoleArn: config.role.roleArn, // required
        LogGroupArn: config.logGroup.logGroupArn, // required
      },
    });
    config.role.grantPassRole(this.customResourceFn.role!);
  }
}

export class KinesisFirehoseEventDestination extends EventDestination {
  readonly deliveryStreamArn: string;
  readonly streamWriterRoleArn: string;
  readonly bucketWriterRoleArn: string;

  constructor(
    scope: Construct,
    id: string,
    config: KinesisFirehoseEventDestinationConfig,
  ) {
    super(scope, id, {
      ConfigurationSetName: config.configurationSetName, // required
      EventDestinationName: config.eventDestinationName, // required
      MatchingEventTypes: config.matchingEventTypes,
      KinesisFirehoseDestination: {
        // CloudWatchLogsDestination
        IamRoleArn: config.streamWriterRole.roleArn, // required
        DeliveryStreamArn: config.destination.deliveryStreamArn, // required
      },
    });
    config.streamWriterRole.grantPassRole(this.customResourceFn.role!);
    this.deliveryStreamArn = config.destination.deliveryStreamArn;
    this.streamWriterRoleArn = config.streamWriterRole.roleArn;
    this.bucketWriterRoleArn = config.bucketWriterRole.roleArn;
  }
}

export class SnsEventDestination extends EventDestination {
  constructor(scope: Construct, id: string, config: SnsEventDestinationConfig) {
    super(scope, id, {
      ConfigurationSetName: config.configurationSetName, // required
      EventDestinationName: config.eventDestinationName, // required
      MatchingEventTypes: config.matchingEventTypes,
      SnsDestination: {
        // CloudWatchLogsDestination
        TopicArn: config.topic.topicArn,
      },
    });
  }
}
