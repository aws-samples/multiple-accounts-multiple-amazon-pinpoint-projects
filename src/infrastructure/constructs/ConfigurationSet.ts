/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this
 * software and associated documentation files (the "Software"), to deal in the Software
 * without restriction, including without limitation the rights to use, copy, modify,
 * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import path from "path";
import { Aws, CustomResource, Duration, RemovalPolicy } from "aws-cdk-lib";
import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Runtime, Tracing } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { Construct, Dependable, IDependable } from "constructs";
import { EventDestination } from "./EventDestination";

export interface ConfigurationSetConfig {
  configurationSetName: string;
}

export class ConfigurationSet extends Construct implements IDependable {
  readonly configurationSetName: string;
  private eventDestinations: { [key: string]: EventDestination | undefined } =
    {};

  constructor(scope: Construct, id: string, config: ConfigurationSetConfig) {
    super(scope, id);
    Dependable.implement(this, {
      dependencyRoots: [this],
    });
    this.node.addValidation({
      validate(): string[] {
        const messages: string[] = [];

        return messages;
      },
    });

    const logGroup = new LogGroup(
      this,
      "ConfigurationSetCustomResourceFnLogGroup",
      {
        retention: RetentionDays.ONE_MONTH,
      },
    );
    const customResourceFn = new NodejsFunction(
      this,
      "ConfigurationSetCustomResourceFn",
      {
        description: "ConfigurationSet CustomResource Function",
        memorySize: 256,
        timeout: Duration.seconds(30),
        runtime: Runtime.NODEJS_LATEST,
        handler: "index.onEvent",
        entry: path.join(
          __dirname,
          "..",
          "..",
          "runtime",
          "ConfigurationSet.ts",
        ),
        logGroup: logGroup,

        environment: {
          LOG_LEVEL: "DEBUG",
        },
        tracing: Tracing.ACTIVE,
        initialPolicy: [
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["sms-voice:*ConfigurationSet"],
            resources: [
              `arn:${Aws.PARTITION}:sms-voice:${Aws.REGION}:${Aws.ACCOUNT_ID}:configuration-set/*`,
            ],
          }),
        ],
      },
    );

    const customResource = new CustomResource(
      this,
      "ConfigurationSetCustomResource",
      {
        serviceToken: customResourceFn.functionArn,
        removalPolicy: RemovalPolicy.DESTROY,
        properties: {
          Name: config.configurationSetName,
        },
      },
    );
    this.configurationSetName = customResource.getAttString(
      "ConfigurationSetName",
    );
  }

  addEventDestination(eventDestination: EventDestination) {
    this.eventDestinations[eventDestination.name] = eventDestination;
    eventDestination.node.addDependency(this);
  }

  getEventDestination(name: string): EventDestination | undefined {
    return this.eventDestinations[name];
  }
}
