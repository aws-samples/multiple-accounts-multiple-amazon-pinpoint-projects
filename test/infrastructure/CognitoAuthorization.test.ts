import {assert, describe, expect, it} from "vitest";
import {CognitoAuthorization} from "../../src/infrastructure/constructs/CognitoAuthorization";
import {Stack} from "aws-cdk-lib";
import {Template} from "aws-cdk-lib/assertions";


describe('CognitoAuthorization', () => {
	it("Must be at least 9 characters long", async () => {
		try {
			const stack = new Stack();
			new CognitoAuthorization(stack, "CognitoAuthorization", {

				apiTempPassword: "12345"
			})
			Template.fromStack(stack);
			assert.isTrue(false, "Should have thrown an exception")
		}catch(e){
			const error=e as Error
			assert.isTrue(error.message.includes("The temp password for the API admin user must be at least 9 characters long, include digits, both lower and uppercase letters as well as symbols"))
		}
	})
	it("Must have digits", async () => {
		try {
			const stack = new Stack();
			new CognitoAuthorization(stack, "CognitoAuthorization", {

				apiTempPassword: "abDgrtsr#"
			})
			Template.fromStack(stack);
			assert.isTrue(false, "Should have thrown an exception")
		}catch(e){
			const error=e as Error
			assert.isTrue(error.message.includes("The temp password for the API admin user must be at least 9 characters long, include digits, both lower and uppercase letters as well as symbols"))
		}
	})
	it("Must have upper case", async () => {
		try {
			const stack = new Stack();
			new CognitoAuthorization(stack, "CognitoAuthorization", {

				apiTempPassword: "ab1grtsr#"
			})
			Template.fromStack(stack);
			assert.isTrue(false, "Should have thrown an exception")
		}catch(e){
			const error=e as Error
			assert.isTrue(error.message.includes("The temp password for the API admin user must be at least 9 characters long, include digits, both lower and uppercase letters as well as symbols"))
		}
	})
	it("Must have lower case", async () => {
		try {
			const stack = new Stack();
			new CognitoAuthorization(stack, "CognitoAuthorization", {

				apiTempPassword: "AB4D#$DAD"
			})
			Template.fromStack(stack);
			assert.isTrue(false, "Should have thrown an exception")
		}catch(e){
			const error=e as Error
			assert.isTrue(error.message.includes("The temp password for the API admin user must be at least 9 characters long, include digits, both lower and uppercase letters as well as symbols"))
		}
	})
	it("Must have special character", async () => {
		try {
			const stack = new Stack();
			new CognitoAuthorization(stack, "CognitoAuthorization", {

				apiTempPassword: "AB4DafDAD"
			})
			Template.fromStack(stack);
			assert.isTrue(false, "Should have thrown an exception")
		}catch(e){
			const error=e as Error
			assert.isTrue(error.message.includes("The temp password for the API admin user must be at least 9 characters long, include digits, both lower and uppercase letters as well as symbols"))
		}
	})
	it("Works when all conditions are met", async () => {

			const stack = new Stack();
			new CognitoAuthorization(stack, "CognitoAuthorization", {

				apiTempPassword: "123$aDCEF"
			})
			const template=Template.fromStack(stack);
			expect(template).toMatchSnapshot()

	})
})