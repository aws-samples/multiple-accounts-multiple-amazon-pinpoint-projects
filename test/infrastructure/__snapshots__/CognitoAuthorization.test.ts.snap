// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CognitoAuthorization > Works when all conditions are met 1`] = `
{
  "Outputs": {
    "ApiAdminTempPassword": {
      "Value": "123$aDCEF",
    },
    "ApiAdminUsername": {
      "Value": "admin",
    },
    "ClientCredentialsClientId": {
      "Value": {
        "Ref": "CognitoAuthorizationUserPoolClientCredentialsClient93F10DFB",
      },
    },
    "ImplicitClientId": {
      "Value": {
        "Ref": "CognitoAuthorizationUserPoolImplicitClient943D320D",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "ed6cd104ff5f101d06dae8cb2b87cc6e6d69b9a22055b467ea6cae10ff023023.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CognitoAuthorizationAdminGroupAA27478A": {
      "Properties": {
        "Description": "Administrator",
        "GroupName": "Admin",
        "UserPoolId": {
          "Ref": "CognitoAuthorizationUserPool4ABAC7A6",
        },
      },
      "Type": "AWS::Cognito::UserPoolGroup",
    },
    "CognitoAuthorizationAdminSetUserPassword520C94C0": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CognitoAuthorizationAdminSetUserPasswordCustomResourcePolicyC5B16053",
        "CognitoAuthorizationAdminUser043054DC",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"CognitoIdentityServiceProvider","action":"adminSetUserPassword","parameters":{"UserPoolId":"",
              {
                "Ref": "CognitoAuthorizationUserPool4ABAC7A6",
              },
              "","Username":"admin","Password":"123$aDCEF","Permanent":false},"physicalResourceId":{"id":"AdminSetUserPassword"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "CognitoAuthorizationAdminSetUserPasswordCustomResourcePolicyC5B16053": {
      "DependsOn": [
        "CognitoAuthorizationAdminUser043054DC",
      ],
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cognito-idp:AdminSetUserPassword",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CognitoAuthorizationAdminSetUserPasswordCustomResourcePolicyC5B16053",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CognitoAuthorizationAdminUser043054DC": {
      "Properties": {
        "UserPoolId": {
          "Ref": "CognitoAuthorizationUserPool4ABAC7A6",
        },
        "Username": "admin",
      },
      "Type": "AWS::Cognito::UserPoolUser",
    },
    "CognitoAuthorizationAdminUserGroupAttachmentBDB815E4": {
      "DependsOn": [
        "CognitoAuthorizationAdminGroupAA27478A",
        "CognitoAuthorizationAdminUser043054DC",
      ],
      "Properties": {
        "GroupName": "Admin",
        "UserPoolId": {
          "Ref": "CognitoAuthorizationUserPool4ABAC7A6",
        },
        "Username": "admin",
      },
      "Type": "AWS::Cognito::UserPoolUserToGroupAttachment",
    },
    "CognitoAuthorizationUserPool4ABAC7A6": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccountRecoverySetting": {
          "RecoveryMechanisms": [
            {
              "Name": "verified_phone_number",
              "Priority": 1,
            },
            {
              "Name": "verified_email",
              "Priority": 2,
            },
          ],
        },
        "AdminCreateUserConfig": {
          "AllowAdminCreateUserOnly": true,
        },
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "Policies": {
          "PasswordPolicy": {
            "MinimumLength": 9,
            "RequireLowercase": true,
            "RequireNumbers": true,
            "RequireSymbols": true,
            "RequireUppercase": true,
            "TemporaryPasswordValidityDays": 1,
          },
        },
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UserPoolAddOns": {
          "AdvancedSecurityMode": "ENFORCED",
        },
        "UserPoolName": "pinpoint-management-account-user-pool",
        "VerificationMessageTemplate": {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Retain",
    },
    "CognitoAuthorizationUserPoolApiResourceServer157544EA": {
      "Properties": {
        "Identifier": "api",
        "Name": "ApiResourceServer",
        "Scopes": [
          {
            "ScopeDescription": "Admin Scope",
            "ScopeName": "Admin",
          },
        ],
        "UserPoolId": {
          "Ref": "CognitoAuthorizationUserPool4ABAC7A6",
        },
      },
      "Type": "AWS::Cognito::UserPoolResourceServer",
    },
    "CognitoAuthorizationUserPoolClientCredentialsClient93F10DFB": {
      "Properties": {
        "AllowedOAuthFlows": [
          "client_credentials",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": [
          {
            "Fn::Join": [
              "",
              [
                {
                  "Ref": "CognitoAuthorizationUserPoolApiResourceServer157544EA",
                },
                "/Admin",
              ],
            ],
          },
        ],
        "CallbackURLs": [
          "https://oauth.pstmn.io/v1/callback",
        ],
        "ClientName": "pinpoint-management-account-user-pool-client-credentials-client",
        "ExplicitAuthFlows": [
          "ALLOW_USER_PASSWORD_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH",
        ],
        "GenerateSecret": true,
        "SupportedIdentityProviders": [
          "COGNITO",
        ],
        "UserPoolId": {
          "Ref": "CognitoAuthorizationUserPool4ABAC7A6",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "CognitoAuthorizationUserPoolCognitoDomain183E3ADA": {
      "Properties": {
        "Domain": {
          "Fn::Join": [
            "",
            [
              "sms-",
              {
                "Ref": "AWS::AccountId",
              },
            ],
          ],
        },
        "UserPoolId": {
          "Ref": "CognitoAuthorizationUserPool4ABAC7A6",
        },
      },
      "Type": "AWS::Cognito::UserPoolDomain",
    },
    "CognitoAuthorizationUserPoolImplicitClient943D320D": {
      "Properties": {
        "AllowedOAuthFlows": [
          "implicit",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": [
          {
            "Fn::Join": [
              "",
              [
                {
                  "Ref": "CognitoAuthorizationUserPoolApiResourceServer157544EA",
                },
                "/Admin",
              ],
            ],
          },
        ],
        "CallbackURLs": [
          "https://oauth.pstmn.io/v1/callback",
        ],
        "ClientName": "pinpoint-management-account-user-pool-implicit-client",
        "ExplicitAuthFlows": [
          "ALLOW_USER_PASSWORD_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH",
        ],
        "SupportedIdentityProviders": [
          "COGNITO",
        ],
        "UserPoolId": {
          "Ref": "CognitoAuthorizationUserPool4ABAC7A6",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
